pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -294,6 +357,132 @@ func contains(values []string, val string) bool {
+		if !ok || v == nil {
+			log.FromContext(ctx).Errorf("invalid result for root certificate watcher")
+			return
+		}
+
+		roots := make([]string, 0, len(v.Roots))
+		for _, root := range v.Roots {
+			roots = append(roots, root.RootCertPEM)
+		}
+
+		dest <- roots
+	}
+}
+
+type keyPair struct {
+	cert string
+	key  string
+}
+
+func leafWatcherHandler(ctx context.Context, dest chan<- keyPair) func(watch.BlockingParamVal, interface{}) {
+	return func(_ watch.BlockingParamVal, raw interface{}) {
+		if raw == nil {
+			return
+		}
+
+		v, ok := raw.(*api.LeafCert)
+		if !ok || v == nil {
+			log.FromContext(ctx).Errorf("invalid result for leaf certificate watcher")
+			return
+		}
+
+		dest <- keyPair{
+			cert: v.CertPEM,
+			key:  v.PrivateKeyPEM,
+		}
+	}
+}
+
+func (p *Provider) createConnectTLSWatchers() (*watch.Plan, *watch.Plan, error) {
+	leafWatcher, err := watch.Parse(map[string]interface{}{
+		"type":    "connect_leaf",
+		"service": p.ServiceName,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create leaf cert watcher plan: %w", err)
+	}
+
+	rootWatcher, err := watch.Parse(map[string]interface{}{
+		"type": "connect_roots",
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create root cert watcher plan: %w", err)
+	}
+
+	return leafWatcher, rootWatcher, nil
+}
+
+func (p *Provider) watchConnectTLS(ctx context.Context, leafWatcher *watch.Plan, rootWatcher *watch.Plan) {
+	ctxLog := log.With(ctx, log.Str(log.ProviderName, "consulcatalog"))
+	logger := log.FromContext(ctxLog)
+
+	leafChan := make(chan keyPair)
+	rootChan := make(chan []string)
+
+	leafWatcher.HybridHandler = leafWatcherHandler(ctx, leafChan)
+	rootWatcher.HybridHandler = rootsWatchHandler(ctx, rootChan)
+
+	logOpts := &hclog.LoggerOptions{
+		Name:       "consulcatalog",
+		Level:      hclog.LevelFromString(logrus.GetLevel().String()),
+		JSONFormat: true,
+	}
+
+	hclogger := hclog.New(logOpts)
+
+	go func() {
+		err := leafWatcher.RunWithClientAndHclog(p.client, hclogger)
+		if err != nil {
+			// This should never be triggered, RunWithClientAndHclog does not seem to ever return anything but nil
+			logger.WithError(err).Errorf("Leaf certificate watcher failed with error")
+		}
+	}()
+
+	go func() {
+		err := rootWatcher.RunWithClientAndHclog(p.client, hclogger)
+		if err != nil {
+			// This should never be triggered, RunWithClientAndHclog does not seem to ever return anything but nil
+			logger.WithError(err).Errorf("Root certificate watcher failed with error")
+		}
+	}()
+
+	var (
+		certInfo  *connectCert
+		leafCerts keyPair
+		rootCerts []string
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			leafWatcher.Stop()
+			rootWatcher.Stop()
+			return
+		case rootCerts = <-rootChan:
+		case leafCerts = <-leafChan:
+		}
+		newCertInfo := &connectCert{
+			service: p.ServiceName,
+			root:    rootCerts,
+			leaf:    leafCerts,
+		}
+		if newCertInfo.isReady() && !reflect.DeepEqual(newCertInfo, certInfo) {
```

Comment:
Should we log or something when we have a new certinfo (different from the previous one), for which isReady() is false? Is that something that can actually happen? If yes, will the situation rectify by itself with subsequent renewals? etc.

Reply:
> Is that something that can actually happen?

Yes, the very first time `rootCerts` and `leafCerts` will not both be ready (or rather `select` will just evaluate one channel) and as such we have a "half-finished" certificate at that point. After that only the `leafCerts` should change regularly. When the `rootCerts` also change (the are valid 10 years usually), then there might be a very small window where the returned cert might be invalid (new root but old leafs). Not sure how to handle that though aside from verifying the leaf against the root chain or similar (I also do not think connect currently handles that case).

Reply:
fair enough.